cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,232 @@
+    on each scale-up operation.
+
+## Proposal
+
+### High level
+
+Provisioning Request (abbr. ProvReq) is a new namespaced Custom Resource that
+aims to allow users to ask CA for capacity for groups of pods.  
+It allows users to express the fact that group of pods is connected and should
+be threated as one entity.  
+This AEP proposes an API that can have multiple provisioning classes and can be
+extended by cloud provider specific ones.  
+This object is meant as one-shot request to CA, so that if CA fails to provision
+the capacity it is up to users to retry (such retry functionality can be added
+later on).
+
+### Provisioning Request CRD
+
+The following code snippets assume [kubebuilder](https://book.kubebuilder.io/)
+is ued to generate the CRD:
+
+```go
+// ProvisioningRequest is a way to express additional capacity
+// that we would like to provision in the cluster. Cluster Autoscaler
+// can use this information in its calculations and signal if the capacity
+// is available in the cluster or actively add capacity if needed.
+type ProvisioningRequest struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	//
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Spec contains specification of the ProvisioningRequest object.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	//
+	// +kubebuilder:validation:Required
+	Spec ProvisioningRequestSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+	// Status of the ProvisioningRequest. CA constantly reconciles this field.
+	//
+	// +optional
+	Status ProvisioningRequestStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ProvisioningRequestList is a object for list of ProvisioningRequest.
+type ProvisioningRequestList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	//
+	// +optional
+	metav1.ListMeta `json:"metadata" protobuf:"bytes,1,opt,name=metadata"`
+	// Items, list of ProvisioningRequest returned from API.
+	//
+	// +optional
+	Items []ProvisioningRequest `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// ProvisioningRequestSpec is a specification of additional pods for which we
+// would like to provision additional resources in the cluster.
+type ProvisioningRequestSpec struct {
+	// PodSets lists groups of pods for which we would like to provision
+	// resources.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	PodSets []PodSet `json:"podSets" protobuf:"bytes,1,rep,name=podSets"`
+
+	// ProvisioningClass describes the different modes of provisioning the resources.
+	// Supported values:
+	// * GENERIC_CHECK_CAPACITY - check if current cluster state can fullfil this request
```

Comment:
I think that there may be quite a few use cases. For example, trying to run low priority jobs only when there is capacity on the statically allocated nodes. Currently K8S doesn't have gang scheduling, and this may work as a best-effort replacement. 
The problem with invalidation exist in all cases where shared/non-exclusive resources are provisioned. This should be addressed in the doc. Possible options are:

* Invalidate request that is still there.
* Treat PR as a point-in-time query and don't bother about invalidation. 

Given that we already have 2 possible behaviors for a single capacity, the params/classes may be actually needed from day 1.